src/marshal.rs

```diff
@@ -0,0 +1,68 @@
+use crate::conversion::AsPyPointer;
+use crate::ffi;
+use crate::{PyErr, PyObject, PyResult, Python};
+use std::os::raw::c_int;
+
+/// The current version of the marshal binary format.
+pub const VERSION: i32 = 4;
+
+/// Serialize an object to bytes using the Python built-in marshal module.
+///
+/// The built-in marshalling only supports a limited range of object.
+/// See the [python documentation](https://docs.python.org/3/library/marshal.html) for more details.
+pub fn dumps(py: Python, object: &PyObject, version: i32) -> PyResult<Vec<u8>> {
+    let bytes = unsafe { ffi::PyMarshal_WriteObjectToString(object.as_ptr(), version as c_int) };
+    if bytes.is_null() {
+        return Err(PyErr::fetch(py));
+    }
+
+    let mut size = 0isize;
+    let mut data = std::ptr::null_mut();
+    unsafe {
+        ffi::PyBytes_AsStringAndSize(bytes, &mut data, &mut size);
+        let data = Vec::from(std::slice::from_raw_parts(data as *const u8, size as usize));
+        ffi::Py_DecRef(bytes);
+        Ok(data)
+    }
+}
+
+/// Deserialize an object from bytes using the Python built-in marshal module.
+pub fn loads(py: Python, data: &impl AsRef<[u8]>) -> PyResult<PyObject> {
```

Comment:
It's better to return `PyResult<&PyAny>`, since it has `downcast*` methods.
`FromPyPointer::from_owned_ptr_or_err` would work here too.